File: app/src/main/java/xyz/lgvalle/chachi/ArticleViewModel.java
package xyz.lgvalle.chachi;

import android.arch.core.util.Function;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.arch.lifecycle.ViewModel;
import android.util.Log;

import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

import xyz.lgvalle.chachi.guardian.Article;
import xyz.lgvalle.chachi.guardian.TheGuardianDataSource;

public class ArticleViewModel extends ViewModel {

    private static final String TAG = ArticleViewModel.class.getSimpleName();
    private LiveData<List<Article>> articles = new MutableLiveData<>();
    private LiveData<Article> selectedItem;
    private TheGuardianDataSource dataSource;
    private String selectedItemId;
    private DatabaseReference myRef;
    private ValueEventListener valueEventListener;

    public ArticleViewModel(TheGuardianDataSource dataSource) {
        this.dataSource = dataSource;
    }

    LiveData<List<Article>> articles() {

        /*
        articles = Transformations.map(dataSource.edition(), new Function<Edition, List<Article>>() {
            @Override
            public List<Article> apply(Edition edition) {
                return edition.getItems();
            }
        });
        */

        Log.d(TAG, "get articles");
        articles = Transformations.map(dataSource.articles(), new Function<List<Article>, List<Article>>() {
            @Override
            public List<Article> apply(List<Article> articles) {
                Log.d(TAG, "get articles: post value");

                return articles;
            }
        });

//        if (articles.getValue() == null) {
//
//            Log.d(TAG, "get articles");
//            final FirebaseDatabase database = FirebaseDatabase.getInstance();
//            myRef = database.getReference("feed/guardian/items");
//
//
//            valueEventListener = new ValueEventListener() {
//                @Override
//                public void onDataChange(DataSnapshot dataSnapshot) {
//
//                    List<Article> items = dataSnapshot.getValue(new GenericTypeIndicator<List<Article>>() {
//                    });
//                    Log.d(TAG, "get articles: post value");
//                    articles.postValue(items);
//                }
//
//                @Override
//                public void onCancelled(DatabaseError databaseError) {
//
//                }
//            };
//
//            myRef.addListenerForSingleValueEvent(valueEventListener);
//        }

        return articles;

    }

    @Override
    protected void onCleared() {
        super.onCleared();
        myRef.removeEventListener(valueEventListener);

    }

    void selectItem(final String selectedItemId) {
        this.selectedItemId = selectedItemId;
    }

    LiveData<Article> selectedItem() {
        selectedItem = Transformations.map(articles, new Function<List<Article>, Article>() {
            @Override
            public Article apply(List<Article> articles) {
                for (Article article : articles) {
                    if (article.getId().equals(selectedItemId)) {
                        return article;
                    }
                }
                return null;
            }
        });
        return selectedItem;
    }


    /**
     * Sentiment value goes from -1.0 to 1.0
     *
     * @param sentimentValue
     */
    public void sentiment(final Integer sentimentValue) {
        List<Article> filteredArticles = new ArrayList<>();
        for (Article article : articles.getValue()) {
            if (article.getScore() >= sentimentValue) {
                filteredArticles.add(article);
            }
        }

//        articles.setValue(filteredArticles);
    }
}

File: app/src/main/java/xyz/lgvalle/chachi/ArticleViewModel.kt
package xyz.lgvalle.chachi

import android.arch.core.util.Function
import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.Transformations
import android.arch.lifecycle.ViewModel
import android.util.Log
import com.google.firebase.database.*
import xyz.lgvalle.chachi.guardian.Article
import xyz.lgvalle.chachi.guardian.TheGuardianDataSource
import java.util.*

class ArticleViewModel(private val dataSource: TheGuardianDataSource) : ViewModel() {
    private var articles: MutableLiveData<List<Article>> = MutableLiveData()
    private var selectedItem: LiveData<Article>? = null
    private var selectedItemId: String? = null
    private val myRef: DatabaseReference? = null
    private val valueEventListener: ValueEventListener? = null

    fun articles(): LiveData<List<Article>> {

        if (articles.value == null) {
            FirebaseDatabase.getInstance()
                    .getReference("feed/articles")
                    .addListenerForSingleValueEvent(object : ValueEventListener {
                        override fun onDataChange(dataSnapshot: DataSnapshot) {
                            if (dataSnapshot.exists()) {
                                val value = dataSnapshot.getValue(object : GenericTypeIndicator<List<Article>>() {

                                })
                                articles.postValue(value)
                            }
                        }

                        override fun onCancelled(databaseError: DatabaseError) {}
                    })
        }

        return articles

    }

    override fun onCleared() {
        super.onCleared()
        myRef!!.removeEventListener(valueEventListener!!)

    }

    internal fun selectItem(selectedItemId: String) {
        this.selectedItemId = selectedItemId
    }

    internal fun selectedItem(): LiveData<Article> {
        selectedItem = Transformations.map(articles, Function { articles ->
            for (article in articles) {
                if (article.id == selectedItemId) {
                    return@Function article
                }
            }
            null
        })
        return selectedItem
    }


    /**
     * Sentiment value goes from -1.0 to 1.0
     *
     * @param sentimentValue
     */
    fun sentiment(sentimentValue: Int?) {
        val filteredArticles = ArrayList<Article>()
        for (article in articles.value!!) {
            if (article.getScore() >= sentimentValue) {
                filteredArticles.add(article)
            }
        }

        //        articles.setValue(filteredArticles);
    }

    companion object {

        private val TAG = ArticleViewModel::class.java.simpleName
    }
}
